cli/azd/pkg/tools/github/github.go

```diff
@@ -323,3 +377,185 @@ func isEnvVarTokenSource(message string) bool {
 
 	return false
 }
+
+func extractFromZip(src, dst string) (string, error) {
+	zipReader, err := zip.OpenReader(src)
+	if err != nil {
+		return "", err
+	}
+
+	defer zipReader.Close()
+
+	var extractedAt string
+	for _, file := range zipReader.File {
+		if !file.FileInfo().IsDir() && strings.Contains(file.Name, "gh") {
+			fileReader, err := file.Open()
+			if err != nil {
+				return extractedAt, err
+			}
+			fileNameParts := strings.Split(file.Name, "/")
+			fileName := fileNameParts[len(fileNameParts)-1]
+			filePath := filepath.Join(dst, fileName)
+			ghCliFile, err := os.OpenFile(filePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, file.Mode())
+			if err != nil {
+				return extractedAt, err
+			}
+			defer ghCliFile.Close()
+			_, err = io.Copy(ghCliFile, fileReader)
+			if err != nil {
+				return extractedAt, err
+			}
+			extractedAt = filePath
+			break
+		}
+	}
+	return extractedAt, nil
+}
+func extractFromTar(src, dst string) (string, error) {
+	gzFile, err := os.Open(src)
+	if err != nil {
+		return "", err
+	}
+	defer gzFile.Close()
+
+	gzReader, err := gzip.NewReader(gzFile)
+	if err != nil {
+		return "", err
+	}
+	defer gzReader.Close()
+
+	var extractedAt string
+	// tarReader doesn't need to be closed as it is closed by the gz reader
+	tarReader := tar.NewReader(gzReader)
+	for {
+		fileHeader, err := tarReader.Next()
+		if errors.Is(err, io.EOF) {
+			return extractedAt, fmt.Errorf("did not find gh cli within tar file: %w", err)
+		}
+		if fileHeader == nil {
+			continue
+		}
+		if err != nil {
+			return extractedAt, err
+		}
+		// Tha name contains the path, remove it
+		fileNameParts := strings.Split(fileHeader.Name, "/")
+		fileName := fileNameParts[len(fileNameParts)-1]
+		if fileHeader.Typeflag == tar.TypeReg && fileName == "gh" {
+			filePath := filepath.Join(dst, fileName)
+			ghCliFile, err := os.OpenFile(filePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, os.FileMode(fileHeader.Mode))
+			if err != nil {
+				return extractedAt, err
+			}
+			defer ghCliFile.Close()
+			_, err = io.Copy(ghCliFile, tarReader)
+			if err != nil {
+				return extractedAt, err
+			}
+			extractedAt = filePath
+			break
+		}
+	}
+
+	return extractedAt, nil
+}
+
+// extractGhCli gets the Github cli from either a zip or a tar.gz
+func extractGhCli(src, dst string) (string, error) {
+	if strings.HasSuffix(src, ".zip") {
+		return extractFromZip(src, dst)
+	} else if strings.HasSuffix(src, ".tar.gz") {
+		return extractFromTar(src, dst)
+	}
+	return "nil", fmt.Errorf("Unknown format")
```

Comment:
Curious... Why return the string with value `nil` instead of `""`? Also, can we codify the return string in the contract?

`(cliPath string, err error)`?

Reply:
I see. It can return a valid path even on `err`. Mind updating docs for `GetGitHubCliImplementation` for this? I feel extremely nervous when methods function differently than the expected standard error-handling behavior..

Reply:
I'll change "nil" to "".  This is a mistake